File: src/app/readiness.rs
//! Independent customer declarations. Evidence contents are never fetched.

use std::collections::BTreeSet;
use std::fmt::Write as _;
use std::fs::File;
use std::io::Read;
use std::path::Path;

use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::{Value, json};

use super::{Exit, write_output};
use crate::AuditError;
use crate::cli::{ReadinessArgs, ReportFormat};

const MAX_BYTES: u64 = 1_048_576;
const VERSION: &str = "canonical.readiness-response/v1";
const NOTICE: &str = "Customer declarations and evidence references only. Evidence contents, operating effectiveness, applicability and reviewer identity have not been verified. This is not certification, attestation, authorization or a legal opinion.";

#[derive(Deserialize)]
struct Catalog {
    version: String,
    frameworks: Vec<Framework>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Framework {
    id: String,
    title: String,
    edition: String,
    scope_note: String,
    sources: Vec<String>,
    questions: Vec<Question>,
}

#[derive(Deserialize)]
struct Question {
    id: String,
    prompt: String,
    evidence: String,
    method: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Context {
    customer_id: String,
    assessment_id: String,
    scope: String,
    period_start: String,
    period_end: String,
    as_of: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Response {
    schema_version: String,
    catalog_version: String,
    framework_id: String,
    context: Context,
    answers: Vec<Answer>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
enum Status {
    Unanswered,
    Implemented,
    Partial,
    Missing,
    NotApplicable,
}

impl Status {
    fn label(self) -> &'static str {
        match self {
            Self::Unanswered => "unanswered",
            Self::Implemented => "implemented",
            Self::Partial => "partial",
            Self::Missing => "missing",
            Self::NotApplicable => "not_applicable",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Answer {
    question_id: String,
    status: Status,
    owner: String,
    notes: String,
    evidence_ref: String,
    evidence_date: String,
    reviewer: String,
    due_date: String,
}

#[derive(Default, Serialize)]
#[serde(rename_all = "camelCase")]
struct Summary {
    total: usize,
    answered: usize,
    declared_gaps: usize,
    unknown_evidence: usize,
    incomplete_metadata: usize,
    review_pending: usize,
    overdue_actions: usize,
}

fn invalid() -> AuditError {
    AuditError::Invalid {
        field: "readiness",
        reason: "invalid packet, framework/version, context, answer, date or input limit".to_owned(),
    }
}

fn decode<T: DeserializeOwned>(bytes: &[u8]) -> Result<T, AuditError> {
    if bytes.len() > 1_048_576 {
        return Err(invalid());
    }
    serde_json::from_slice(bytes).map_err(|_| invalid())
}

fn read<T: DeserializeOwned>(path: &Path) -> Result<T, AuditError> {
    let file = File::open(path)?;
    if !file.metadata()?.is_file() {
        return Err(invalid());
    }
    let mut bytes = Vec::new();
    file.take(MAX_BYTES + 1).read_to_end(&mut bytes)?;
    decode(&bytes)
}

fn identifier(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 128
        && value.as_bytes()[0].is_ascii_alphanumeric()
        && value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || b"._:-".contains(&byte))
}

fn text(value: &str, maximum: usize) -> bool {
    value.chars().count() <= maximum
        && !value
            .chars()
            .any(|ch| ch.is_ascii_control() && !matches!(ch, '\t' | '\n' | '\r'))
}

// Match ECMAScript String.trim rather than Rust's different Unicode whitespace set.
fn blank_text(value: &str) -> bool {
    value.chars().all(|ch| {
        matches!(
            ch,
            '\u{0009}'..='\u{000d}'
                | '\u{0020}'
                | '\u{00a0}'
                | '\u{1680}'
                | '\u{2000}'..='\u{200a}'
                | '\u{2028}'
                | '\u{2029}'
                | '\u{202f}'
                | '\u{205f}'
                | '\u{3000}'
                | '\u{feff}'
        )
    })
}

fn valid_date(value: &str) -> bool {
    let bytes = value.as_bytes();
    if bytes.len() != 10
        || bytes[4] != b'-'
        || bytes[7] != b'-'
        || bytes
            .iter()
            .enumerate()
            .any(|(index, byte)| !matches!(index, 4 | 7) && !byte.is_ascii_digit())
    {
        return false;
    }
    let number = |start: usize, end: usize| {
        bytes[start..end]
            .iter()
            .fold(0_u32, |value, byte| value * 10 + u32::from(byte - b'0'))
    };
    let year = number(0, 4);
    let month = number(5, 7);
    let day = number(8, 10);
    let leap = year.is_multiple_of(4) && (!year.is_multiple_of(100) || year.is_multiple_of(400));
    let maximum = match month {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
        2 if leap => 29,
        2 => 28,
        _ => 0,
    };
    year >= 1900 && day > 0 && day <= maximum
}

fn validate_context(context: &Context) -> Result<(), AuditError> {
    if !identifier(&context.customer_id)
        || !identifier(&context.assessment_id)
        || !text(&context.scope, 1000)
        || blank_text(&context.scope)
        || !valid_date(&context.period_start)
        || !valid_date(&context.period_end)
        || !valid_date(&context.as_of)
        || context.period_start > context.period_end
        || context.period_end > context.as_of
    {
        return Err(invalid());
    }
    Ok(())
}

fn blank(question: &Question) -> Answer {
    Answer {
        question_id: question.id.clone(),
        status: Status::Unanswered,
        owner: String::new(),
        notes: String::new(),
        evidence_ref: String::new(),
        evidence_date: String::new(),
        reviewer: String::new(),
        due_date: String::new(),
    }
}

fn draft(catalog: &Catalog, framework: &Framework, context: &Context) -> Response {
    Response {
        schema_version: VERSION.to_owned(),
        catalog_version: catalog.version.clone(),
        framework_id: framework.id.clone(),
        context: context.clone(),
        answers: framework.questions.iter().map(blank).collect(),
    }
}

fn validate_answer(answer: &Answer, context: &Context) -> bool {
    text(&answer.owner, 128)
        && text(&answer.reviewer, 128)
        && text(&answer.notes, 4000)
        && (answer.evidence_ref.is_empty() || identifier(&answer.evidence_ref))
        && (answer.evidence_date.is_empty() || valid_date(&answer.evidence_date))
        && answer.evidence_date <= context.as_of
        && (answer.due_date.is_empty() || valid_date(&answer.due_date))
}

fn validate(
    catalog: &Catalog,
    framework: &Framework,
    context: &Context,
    response: &Response,
) -> Result<(), AuditError> {
    validate_context(context)?;
    if response.schema_version != VERSION
        || response.catalog_version != catalog.version
        || response.framework_id != framework.id
        || response.context != *context
        || response.answers.len() > framework.questions.len()
    {
        return Err(invalid());
    }
    let known: BTreeSet<_> = framework.questions.iter().map(|q| &q.id).collect();
    let mut seen = BTreeSet::new();
    for answer in &response.answers {
        if !known.contains(&answer.question_id)
            || !seen.insert(&answer.question_id)
            || !validate_answer(answer, context)
        {
            return Err(invalid());
        }
    }
    Ok(())
}

fn answer_issues(answer: &Answer, as_of: &str, summary: &mut Summary) -> Vec<&'static str> {
    let mut issues = Vec::new();
    if answer.status == Status::Unanswered {
        issues.push("unanswered");
        return issues;
    }
    summary.answered += 1;
    if blank_text(&answer.owner) || blank_text(&answer.notes) {
        issues.push("owner_or_explanation_missing");
    }
    if blank_text(&answer.reviewer) {
        issues.push("assessor_review_pending");
        summary.review_pending += 1;
    }
    if answer.status == Status::Implemented
        && (answer.evidence_ref.is_empty() || answer.evidence_date.is_empty())
    {
        issues.push("evidence_unknown");
        summary.unknown_evidence += 1;
    }
    if matches!(answer.status, Status::Partial | Status::Missing) {
        summary.declared_gaps += 1;
        issues.push("declared_gap");
        if answer.due_date.is_empty() {
            issues.push("remediation_date_missing");
        } else if answer.due_date.as_str() < as_of {
            issues.push("remediation_overdue");
            summary.overdue_actions += 1;
        }
    }
    if answer.status == Status::NotApplicable && blank_text(&answer.reviewer) {
        issues.push("applicability_review_missing");
    }
    if issues.iter().any(|issue| {
        matches!(
            *issue,
            "owner_or_explanation_missing"
                | "remediation_date_missing"
                | "applicability_review_missing"
        )
    }) {
        summary.incomplete_metadata += 1;
    }
    issues
}

fn analyze(framework: &Framework, response: &Response) -> (Value, bool) {
    let mut summary = Summary::default();
    let mut items = Vec::new();
    for question in &framework.questions {
        summary.total += 1;
        let missing = blank(question);
        let answer = response
            .answers
            .iter()
            .find(|answer| answer.question_id == question.id)
            .unwrap_or(&missing);
        let issues = answer_issues(answer, &response.context.as_of, &mut summary);
        items.push(json!({"questionId": question.id, "status": answer.status, "issues": issues}));
    }
    let gaps = summary.answered < summary.total
        || summary.declared_gaps > 0
        || summary.unknown_evidence > 0
        || summary.incomplete_metadata > 0;
    let report = json!({
        "schemaVersion": "canonical.readiness-report/v1",
        "catalogVersion": response.catalog_version,
        "frameworkId": framework.id,
        "context": response.context,
        "summary": summary,
        "items": items,
        "declaredGapsOrUnknowns": gaps,
        "assurance": "none",
        "notice": NOTICE
    });
    (report, gaps)
}

fn escaped(value: &str) -> String {
    let mut result = String::new();
    for ch in value.chars() {
        match ch {
            '&' => result.push_str("&amp;"),
            '<' => result.push_str("&lt;"),
            '>' => result.push_str("&gt;"),
            '\n' | '\r' | '\t' => result.push(' '),
            ch if "\\`*_{}[]()#+.!|~-".contains(ch) => {
                result.push('\\');
                result.push(ch);
            }
            ch => result.push(ch),
        }
    }
    result
}

fn markdown(framework: &Framework, response: &Response, report: &Value) -> String {
    let mut output = format!(
        "# {} readiness checklist\n\nEdition: {}. Catalog: {}.\n\n{}\n\n{}\n\nComplete this framework independently. Missing evidence is unknown, not proof of failure.\n\n",
        framework.title, framework.edition, response.catalog_version, framework.scope_note, NOTICE
    );
    let context = &response.context;
    for (name, value) in [
        ("customerId", &context.customer_id),
        ("assessmentId", &context.assessment_id),
        ("scope", &context.scope),
        ("periodStart", &context.period_start),
        ("periodEnd", &context.period_end),
        ("asOf", &context.as_of),
    ] {
        let _ = writeln!(output, "{name}: {}\n", escaped(value));
    }
    let _ = writeln!(output, "Summary: {}\n", report["summary"]);
    for question in &framework.questions {
        let missing = blank(question);
        let answer = response
            .answers
            .iter()
            .find(|answer| answer.question_id == question.id)
            .unwrap_or(&missing);
        let _ = writeln!(
            output,
            "## {} — {}\n\nSuggested evidence: {}\n\nPrimary review method: {}.\n",
            question.id, question.prompt, question.evidence, question.method
        );
        for (name, value) in [
            ("status", answer.status.label()),
            ("owner", answer.owner.as_str()),
            ("notes", answer.notes.as_str()),
            ("evidenceRef", answer.evidence_ref.as_str()),
            ("evidenceDate", answer.evidence_date.as_str()),
            ("reviewer", answer.reviewer.as_str()),
            ("dueDate", answer.due_date.as_str()),
        ] {
            let _ = writeln!(output, "{name}: {}\n", escaped(value));
        }
    }
    output.push_str("## Authoritative references\n\n");
    for source in &framework.sources {
        let _ = writeln!(output, "- {source}");
    }
    output
}

pub(super) fn execute(arguments: &ReadinessArgs) -> Result<Exit, AuditError> {
    let catalog: Catalog = serde_json::from_str(include_str!("../../readiness/catalog.json"))?;
    let framework = catalog
        .frameworks
        .iter()
        .find(|framework| framework.id == arguments.framework)
        .ok_or_else(invalid)?;
    let context: Context = read(&arguments.context)?;
    validate_context(&context)?;
    let response: Response = match &arguments.responses {
        Some(path) => read(path)?,
        None => draft(&catalog, framework, &context),
    };
    validate(&catalog, framework, &context, &response)?;
    let (report, gaps) = analyze(framework, &response);
    let output = match arguments.format {
        ReportFormat::Json if arguments.responses.is_none() => {
            format!("{}\n", serde_json::to_string_pretty(&response)?)
        }
        ReportFormat::Json => format!("{}\n", serde_json::to_string_pretty(&report)?),
        ReportFormat::Markdown => markdown(framework, &response, &report),
    };
    write_output(&arguments.output, &output)?;
    Ok(if arguments.responses.is_some() && gaps {
        Exit::FindingThreshold
    } else {
        Exit::Success
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shared_javascript_fixtures_have_identical_semantics() -> Result<(), AuditError> {
        let catalog: Catalog = serde_json::from_str(include_str!("../../readiness/catalog.json"))?;
        let context: Context = decode(include_bytes!("../../readiness/context.example.json"))?;
        let fixtures: Value = serde_json::from_str(include_str!("../../readiness/fixtures.json"))?;
        let cases = fixtures.as_array().ok_or_else(invalid)?;
        for case in cases {
            let response: Response = serde_json::from_value(case["response"].clone())?;
            let framework = catalog
                .frameworks
                .iter()
                .find(|item| item.id == response.framework_id)
                .ok_or_else(invalid)?;
            validate(&catalog, framework, &context, &response)?;
            assert_eq!(analyze(framework, &response).0, case["report"]);
        }
        Ok(())
    }

    #[test]
    fn every_framework_is_independent_and_inputs_fail_closed() -> Result<(), AuditError> {
        let catalog: Catalog = serde_json::from_str(include_str!("../../readiness/catalog.json"))?;
        let context: Context = decode(include_bytes!("../../readiness/context.example.json"))?;
        assert_eq!(catalog.frameworks.len(), 15);
        for framework in &catalog.frameworks {
            let response = draft(&catalog, framework, &context);
            validate(&catalog, framework, &context, &response)?;
            assert_eq!(response.answers.len(), 10);
            let (report, gaps) = analyze(framework, &response);
            assert!(gaps);
            assert_eq!(report["assurance"], "none");
            let mut wrong = response.clone();
            wrong.context.customer_id = "other-customer".to_owned();
            assert!(validate(&catalog, framework, &context, &wrong).is_err());
            wrong = response.clone();
            wrong.framework_id = "other-framework".to_owned();
            assert!(validate(&catalog, framework, &context, &wrong).is_err());
            wrong = response.clone();
            wrong.answers[1].question_id.clone_from(&response.answers[0].question_id);
            assert!(validate(&catalog, framework, &context, &wrong).is_err());
            wrong = response;
            wrong.answers[0].evidence_ref = "https://example.invalid/private?credential=x".to_owned();
            assert!(validate(&catalog, framework, &context, &wrong).is_err());
        }
        Ok(())
    }

    #[test]
    fn dates_and_markdown_are_fail_closed() {
        for value in ["2026-02-29", "1899-12-31", "2026-13-01", "2026-01-00", "2026-1-01"] {
            assert!(!valid_date(value));
        }
        assert!(blank_text("\u{feff}"));
        assert!(!blank_text("\u{0085}"));
        assert!(valid_date("2024-02-29"));
        assert!(!valid_date("1900-02-29"));
        assert!(valid_date("2000-02-29"));
        assert_eq!(escaped("<img>\n![x](y)"), "&lt;img&gt; \\!\\[x\\]\\(y\\)");
    }

    #[test]
    fn duplicate_fields_and_unpaired_surrogates_are_rejected() {
        assert!(decode::<Context>(br#"{"customerId":"x","customerId":"y"}"#).is_err());
        assert!(decode::<String>(br#""\ud800""#).is_err());
        assert!(decode::<Value>(&vec![b' '; 1_048_577]).is_err());
    }
}

File: readiness/readiness.mjs
// Shared, side-effect-free readiness contract. No persistence, network or inheritance.
export const RESPONSE_VERSION = 'canonical.readiness-response/v1';
export const MAX_BYTES = 1048576;
export const STATUSES = Object.freeze(['unanswered', 'implemented', 'partial', 'missing', 'not_applicable']);
const CONTEXT_KEYS = ['customerId', 'assessmentId', 'scope', 'periodStart', 'periodEnd', 'asOf'];
const ANSWER_KEYS = ['questionId', 'status', 'owner', 'notes', 'evidenceRef', 'evidenceDate', 'reviewer', 'dueDate'];
const ID = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
function object(value, keys, label) {
  if (!value || typeof value !== 'object' || Array.isArray(value) || Object.keys(value).length !== keys.length || keys.some((key) => !Object.hasOwn(value, key))) throw new Error(`Invalid ${label} fields`);
}
function text(value, maximum, label, required = false) {
  if (typeof value !== 'string' || /[\uD800-\uDFFF]/u.test(value) || [...value].length > maximum || /[\u0000-\u0008\u000b\u000c\u000e-\u001f\u007f]/.test(value) || (required && !value.trim())) throw new Error(`Invalid ${label}`);
}
export function validDate(value) {
  if (typeof value !== 'string' || !/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/.test(value) || value < '1900-01-01' || value > '9999-12-31') return false;
  const date = new Date(`${value}T00:00:00Z`);
  return Number.isFinite(date.getTime()) && date.toISOString().slice(0, 10) === value;
}
export function validateContext(context) {
  object(context, CONTEXT_KEYS, 'context');
  if (!ID.test(context.customerId) || !ID.test(context.assessmentId) || typeof context.customerId !== 'string' || typeof context.assessmentId !== 'string') throw new Error('Invalid customer or assessment identifier');
  text(context.scope, 1000, 'scope', true);
  for (const key of ['periodStart', 'periodEnd', 'asOf']) if (!validDate(context[key])) throw new Error('Invalid assessment date');
  if (context.periodStart > context.periodEnd || context.periodEnd > context.asOf) throw new Error('Evidence period must end on or before the as-of date');
}
export function framework(catalog, id) {
  const found = catalog.frameworks.find((item) => item.id === id);
  if (!found) throw new Error('Unknown framework');
  return found;
}
export function createDraft(catalog, id, context) {
  validateContext(context);
  return {
    schemaVersion: RESPONSE_VERSION, catalogVersion: catalog.version, frameworkId: id, context: { ...context },
    answers: framework(catalog, id).questions.map((question) => ({ questionId: question.id, status: 'unanswered', owner: '', notes: '', evidenceRef: '', evidenceDate: '', reviewer: '', dueDate: '' })),
  };
}
export function validate(catalog, id, context, draft) {
  validateContext(context);
  object(draft, ['schemaVersion', 'catalogVersion', 'frameworkId', 'context', 'answers'], 'response');
  if (draft.schemaVersion !== RESPONSE_VERSION || draft.catalogVersion !== catalog.version || draft.frameworkId !== id) throw new Error('Framework or version mismatch; answers cannot be inherited');
  validateContext(draft.context);
  if (CONTEXT_KEYS.some((key) => draft.context[key] !== context[key])) throw new Error('Customer, assessment, scope or evidence-period mismatch');
  const questions = framework(catalog, id).questions, known = new Set(questions.map((item) => item.id)), seen = new Set();
  if (!Array.isArray(draft.answers) || draft.answers.length > questions.length) throw new Error('Invalid answer collection');
  for (const answer of draft.answers) {
    object(answer, ANSWER_KEYS, 'answer');
    if (!known.has(answer.questionId) || seen.has(answer.questionId)) throw new Error('Unknown or duplicate question identifier');
    seen.add(answer.questionId);
    if (!STATUSES.includes(answer.status)) throw new Error('Invalid answer status');
    for (const key of ['owner', 'reviewer']) text(answer[key], 128, key);
    text(answer.notes, 4000, 'notes'); text(answer.evidenceRef, 128, 'evidence reference');
    if (answer.evidenceRef && !ID.test(answer.evidenceRef)) throw new Error('Use an opaque evidence identifier, not a URL or evidence contents');
    for (const key of ['evidenceDate', 'dueDate']) if (answer[key] !== '' && !validDate(answer[key])) throw new Error('Invalid evidence or remediation date');
    if (answer.evidenceDate > context.asOf) throw new Error('Evidence cannot be collected after the as-of date');
  }
  return draft;
}
export function parseResponse(raw) {
  if (typeof raw !== 'string' || new TextEncoder().encode(raw).length > MAX_BYTES) throw new Error('Response exceeds the 1 MiB limit');
  let parsed;
  try { parsed = JSON.parse(raw); } catch { throw new Error('Invalid response JSON'); }
  // Reject ambiguous duplicate object keys, including differently escaped keys.
  // JSON.parse alone silently keeps the last occurrence; Rust serde rejects it.
  const tokens = raw.match(/"(?:\\[\s\S]|[^"\\])*"|[{}\[\]:,]|-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?|true|false|null/g) ?? [];
  let index = 0;
  function walk(depth) {
    if (depth > 32) throw new Error('Response nesting exceeds the limit');
    const token = tokens[index++];
    if (token === '{') {
      const keys = new Set();
      if (tokens[index] === '}') { index++; return; }
      while (true) {
        const key = JSON.parse(tokens[index++]);
        if (keys.has(key)) throw new Error('Duplicate JSON field');
        keys.add(key); index++; walk(depth + 1);
        if (tokens[index++] === '}') return;
      }
    }
    if (token === '[') {
      if (tokens[index] === ']') { index++; return; }
      while (true) { walk(depth + 1); if (tokens[index++] === ']') return; }
    }
  }
  walk(0);
  if (index !== tokens.length) throw new Error('Invalid response JSON');
  return parsed;
}
export function importDraft(catalog, id, context, raw) {
  const draft = parseResponse(raw);
  validate(catalog, id, context, draft);
  // Copy only the validated contract; missing answers remain explicitly unanswered.
  const result = createDraft(catalog, id, context), incoming = new Map(draft.answers.map((answer) => [answer.questionId, answer]));
  result.answers = result.answers.map((answer) => ({ ...(incoming.get(answer.questionId) ?? answer) }));
  return result;
}
export function analyze(catalog, id, context, draft) {
  validate(catalog, id, context, draft);
  const answers = new Map(draft.answers.map((answer) => [answer.questionId, answer]));
  const summary = { total: 0, answered: 0, declaredGaps: 0, unknownEvidence: 0, incompleteMetadata: 0, reviewPending: 0, overdueActions: 0 };
  const items = framework(catalog, id).questions.map((question) => {
    const answer = answers.get(question.id), status = answer?.status ?? 'unanswered';
    const issues = [];
    summary.total++;
    if (status === 'unanswered') issues.push('unanswered');
    else {
      summary.answered++;
      if (!answer.owner.trim() || !answer.notes.trim()) issues.push('owner_or_explanation_missing');
      if (!answer.reviewer.trim()) { issues.push('assessor_review_pending'); summary.reviewPending++; }
      if (status === 'implemented' && (!answer.evidenceRef || !answer.evidenceDate)) { issues.push('evidence_unknown'); summary.unknownEvidence++; }
      if (status === 'partial' || status === 'missing') {
        summary.declaredGaps++; issues.push('declared_gap');
        if (!answer.dueDate) issues.push('remediation_date_missing');
        else if (answer.dueDate < context.asOf) { issues.push('remediation_overdue'); summary.overdueActions++; }
      }
      if (status === 'not_applicable' && !answer.reviewer.trim()) issues.push('applicability_review_missing');
    }
    if (issues.some((issue) => ['owner_or_explanation_missing', 'remediation_date_missing', 'applicability_review_missing'].includes(issue))) summary.incompleteMetadata++;
    return { questionId: question.id, status, issues };
  });
  return {
    schemaVersion: 'canonical.readiness-report/v1', catalogVersion: catalog.version, frameworkId: id, context: { ...context }, summary, items,
    declaredGapsOrUnknowns: summary.answered < summary.total || summary.declaredGaps > 0 || summary.unknownEvidence > 0 || summary.incompleteMetadata > 0,
    assurance: 'none',
    notice: 'Customer declarations and evidence references only. Evidence contents, operating effectiveness, applicability and reviewer identity have not been verified. This is not certification, attestation, authorization or a legal opinion.',
  };
}
function markdownText(value) {
  return String(value).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/([\\`*_{}\[\]()#+.!|~-])/g, '\\$1').replace(/[\r\n\t]/g, ' ');
}
export function markdown(catalog, id, context, draft) {
  const result = analyze(catalog, id, context, draft), selected = framework(catalog, id);
  const answers = new Map(draft.answers.map((answer) => [answer.questionId, answer]));
  const lines = [`# ${selected.title} readiness checklist`, '', `Edition: ${selected.edition}. Catalog: ${catalog.version}.`, '', selected.scopeNote, '', result.notice, '', 'Complete this framework independently. Missing evidence is unknown, not proof of failure. Do not paste secrets, PHI, personal records or raw evidence here.', ''];
  for (const key of CONTEXT_KEYS) lines.push(`${key}: ${markdownText(context[key])}`, '');
  lines.push(`Answered: ${result.summary.answered}/${result.summary.total}. Declared gaps: ${result.summary.declaredGaps}. Unknown evidence: ${result.summary.unknownEvidence}. Incomplete metadata: ${result.summary.incompleteMetadata}. Review pending: ${result.summary.reviewPending}.`, '');
  for (const question of selected.questions) {
    const answer = answers.get(question.id);
    lines.push(`## ${question.id} — ${question.prompt}`, '', `Suggested evidence: ${question.evidence}`, '', `Primary review method: ${question.method}.`, '');
    for (const key of ANSWER_KEYS.filter((key) => key !== 'questionId')) lines.push(`${key}: ${markdownText(answer?.[key] ?? (key === 'status' ? 'unanswered' : ''))}`, '');
  }
  lines.push('## Authoritative references', '', ...selected.sources.map((url) => `- ${url}`), '');
  return lines.join('\n');
}

// A draft can contain valuable evidence/ownership work before its status changes.
export function hasAnswerEdits(draft) {
  return draft.answers.some((answer) => ANSWER_KEYS.some((key) =>
    key !== 'questionId' && answer[key] !== (key === 'status' ? 'unanswered' : '')));
}

// Only one file read/replacement may be active. Failed or cancelled operations
// release the gate; ignored overlapping calls cannot reset an active operation.
export function exclusiveOperation() {
  let active = false;
  return async (operation) => {
    if (active) return false;
    active = true;
    try { await operation(); return true; }
    finally { active = false; }
  };
}

export const QUESTION_VIEWS = Object.freeze(['all', 'unanswered', 'gaps', 'evidence', 'review', 'overdue']);
export function visibleQuestionIds(report, view) {
  if (!QUESTION_VIEWS.includes(view)) throw new Error('Unknown question view');
  const issue = { unanswered: 'unanswered', gaps: 'declared_gap', evidence: 'evidence_unknown', review: 'assessor_review_pending', overdue: 'remediation_overdue' }[view];
  return new Set(report.items.filter((item) => view === 'all' || item.issues.includes(issue)).map((item) => item.questionId));
}

// Planning template only: no customer declaration becomes an assessor result.
// The existing response contract remains the authoritative input boundary.
export function workpaper(catalog, id, context, draft) {
  const report = analyze(catalog, id, context, draft), selected = framework(catalog, id);
  const answers = new Map(draft.answers.map((answer) => [answer.questionId, answer]));
  const lines = [`# ${markdownText(selected.title)} pre-audit workpaper`, '',
    `Edition: ${markdownText(selected.edition)}. Catalog: ${markdownText(catalog.version)}.`, '',
    'Planning template; no assessment has been performed by this export. All assessor outcomes start as not assessed. Customer declarations, N/A decisions and reviewer names are not inherited as assurance.', '',
    report.notice, '', markdownText(selected.scopeNote), ''];
  for (const key of CONTEXT_KEYS) lines.push(`${key}: ${markdownText(context[key])}`, '');
  lines.push('## Engagement authorization and coverage', '',
    'Lead assessor / competence / independence: ____________________', '',
    'Applicable licensed/public requirement program and edition: ____________________', '',
    'Scope exclusions and approved applicability rationale: ____________________', '',
    'Written test authorization, allowed targets/operations, window and stop contact: ____________________', '',
    'Evidence repository access, minimization, retention and deletion agreement: ____________________', '',
    'These broad intake prompts must be expanded against the applicable full requirement program. Choose depth, coverage and sampling for this engagement; no universal sample count is implied.', '');
  for (const question of selected.questions) {
    const answer = answers.get(question.id);
    lines.push(`## Workpaper ${markdownText(question.id)}`, '',
      `Intake objective: ${markdownText(question.prompt)}`, '',
      `Customer declaration (unverified): ${markdownText(answer?.status ?? 'unanswered')}`, '',
      `Customer owner (unverified): ${markdownText(answer?.owner ?? '')}`, '',
      `Evidence request: ${markdownText(question.evidence)}`, '',
      `Existing opaque reference (unverified): ${markdownText(answer?.evidenceRef ?? '')}`, '',
      `Suggested primary method: ${markdownText(question.method)}; assessor confirms examine / interview / test and complementary methods.`, '',
      'Requirement reference / applicability / testable determination: ____________________', '',
      'Evidence-request owner / due date / observation period: ____________________', '',
      'Population definition / source query / size / completeness reconciliation: ____________________', '',
      'Selection method / risk strata / sample-size rationale / selected opaque item IDs: ____________________', '',
      'Zero-event population, substitutions and limitations: ____________________', '',
      'Authorized test steps / expected behavior / actual observation: ____________________', '',
      'Evidence IDs / collector / collected-at / tool version / integrity digest / redactions: ____________________', '',
      'Design evaluation / operating-period evidence / contradictions: ____________________', '',
      'Assessor outcome: not assessed', '',
      'Finding ID / risk / remediation owner / target date / retest criteria: ____________________', '',
      'Retest evidence and residual limitations: ____________________', '',
      'Independent review: pending; authenticated approval reference: ____________________', '');
  }
  lines.push('## Handoff', '',
    'Unresolved requests / exclusions / residual uncertainty: ____________________', '',
    'Qualified assessor recommendation and approval reference: ____________________', '',
    'Store completed workpapers in approved protected storage. Do not paste credentials, PHI or raw personal records. Completing this template is not certification or a legal opinion.', '',
    '## Methodology reference', '',
    'NIST assessment planning guidance (adapt to the selected framework; not a cross-framework control mapping): https://csrc.nist.gov/pubs/sp/800/53/a/r5/final', '');
  return lines.join('\n');
}
